src/evmIndexer/utils/index.ts

```diff
@@ -0,0 +1,326 @@
+    fee: await getFee(event, fromDomain, provider),
+  };
+}
+
+export function parseDestination(
+  hexData: string,
+  domain: Domain,
+  resourceType: DepositType,
+  substrateRpcUrlConfig: Map<number, ApiPromise>,
+): string {
+  const arrayifyData = decodeHex(hexData);
+  let recipient = "";
+  switch (resourceType) {
+    case DepositType.FUNGIBLE:
+    case DepositType.NONFUNGIBLE: {
+      const recipientlen = Number(
+        "0x" + arrayifyData.subarray(32, 64).toString("hex"),
+      );
+      recipient =
+        "0x" + arrayifyData.subarray(64, 64 + recipientlen).toString("hex");
+      break;
+    }
+    case DepositType.PERMISSIONLESS_GENERIC:
+      {
+        // 32 + 2 + 1 + 1 + 20 + 20
+        const lenExecuteFuncSignature = Number(
+          "0x" + arrayifyData.subarray(32, 34).toString("hex"),
+        );
+        const lenExecuteContractAddress = Number(
+          "0x" +
+            arrayifyData
+              .subarray(
+                34 + lenExecuteFuncSignature,
+                35 + lenExecuteFuncSignature,
+              )
+              .toString("hex"),
+        );
+        recipient =
+          "0x" +
+          arrayifyData
+            .subarray(
+              35 + lenExecuteFuncSignature,
+              35 + lenExecuteFuncSignature + lenExecuteContractAddress,
+            )
+            .toString("hex");
+      }
+      break;
+    default:
+      logger.error(`Unsupported resource type: ${resourceType}`);
+      return "";
+  }
+
+  let destination = "";
+  switch (domain.type) {
+    case Network.EVM:
+      destination = recipient;
+      break;
+    case Network.SUBSTRATE: {
+      const substrateAPI = substrateRpcUrlConfig.get(domain.id);
+      if (!substrateAPI) {
+        throw new Error(
+          `Substrate domain with id ${domain.id} not found in RPC configuration variable`,
+        );
+      }
+      destination = parseSubstrateDestination(recipient, substrateAPI);
+      break;
+    }
+  }
+  return destination;
+}
+
+function parseSubstrateDestination(
+  recipient: string,
+  substrateAPI: ApiPromise,
+): string {
+  const decodedData = substrateAPI.createType("MultiLocation", recipient);
+  const multiAddress = decodedData.toJSON() as unknown as MultiLocation;
+  for (const [, junctions] of Object.entries(multiAddress.interior)) {
+    const junston = junctions as Junction;
+    if (junston.accountId32?.id) {
+      return junston.accountId32.id;
+    }
+  }
+  return "";
+}
+
+export function decodeAmountsOrTokenId(
+  data: string,
+  decimals: number,
+  resourceType: DepositType,
+): string | Error {
+  switch (resourceType) {
+    case DepositType.FUNGIBLE: {
+      const amount = AbiCoder.defaultAbiCoder().decode(
+        ["uint256"],
+        data,
+      )[0] as BigNumberish;
+      return formatUnits(amount, decimals);
+    }
+    case DepositType.NONFUNGIBLE: {
+      const tokenId = AbiCoder.defaultAbiCoder().decode(
+        ["uint256"],
+        data,
+      )[0] as number;
```

Comment:
```suggestion
      )[0] as bigint;
```
I don't think it's safe to assume, it will always be number